PrestaShop/PrestaShop
# Enable addCss and addJs for migrated pages and ControllerStack

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop (I need to rebase)
| Description?  | Enable the use of `addCSS()` and `addJS()` in migrated pages, introduces ControllerStack
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes partially https://github.com/PrestaShop/PrestaShop/issues/13943
| How to test?  | See below

# How to test

Install module 
[demotestaddcss.zip](https://github.com/PrestaShop/PrestaShop/files/5232476/demotestaddcss.zip)

Open "customer" listing BO page, a "It works!" alert appears because the module successfully injected `myscript.css` inside the page.

# PR content

## addCss & friends

This PR enables the use of `addCSS()`, `addJS()`, `removeCSS()` and `removeJS()` in Symfony controllers. It was a very useful function available for legacy controllers.

So now you can do `$customerController->addJS($myJavascriptFile);` to inject a JS file into Customer BO page. That is great! 🎉 

## Usage inside hooks

If you are using a Core controller, it's very easy to use `addCSS()` or `addJS()`. Same if you are using a module Symfony controller.

**But a second issues comes:** it is actually hard to use `addCss` or `addJs` **in a hook** ⛵ !

See the following example (from the sample module)
```
public function hookActionCustomerGridDefinitionModifier(array $params)
    {
        
    }
```
You are inside this hook, which goal is to modify the "Customer Grid". Let's say you want to add a new column AND the JS file in charge of managing this column (for example the new column uses a specific ColorPicker). How do you inject the JS file using `addJS()`

What you need to do is:
- explore the code to find out that this hook is fired when user browses "Customer Grid BO page", which is powered by CustomerController
- retrieve the `$customerController` instance

Now that you know the Controller that you want to target is CustomerController, you can do this
```
    public function hookActionCustomerGridDefinitionModifier(array $params)
        $controller = $this->get('PrestaShopBundle\Controller\Admin\Sell\Customer\CustomerController');
        $controller->addJS('myscript.js');
```

## Hooks fired from multiple places

On this hook we are lucky, the hook will be triggered only on Customer Grid page. So we know 100% "the controller that we need to retrieve is CustomerController". **However there are hooks that can be fired from multiple BO pages.**

**So we need a way to know "I am plugged on hook Foo, what is the Controller that handled the HTTP request ?"

In a hook function, the hook params contain PS version, route name, request, but not controller. From request it is possible to find the controller name but it's very dirty.

**Second issue**: even you can find the controller name, but to retrieve the controller _instance_ is even more complex.

➡️ So I did this

### Use controller stack

I introduced a ControllerStack whose behavior is similar to Symfony RequestStack. It listens to Symfony kernel and registers the controllers being matched to HTTP requests.

This allows you to retrieve the "current controller".

Example:
```
    public function hookActionCustomerGridDefinitionModifier(array $params)

        $controller = $this->get('prestashop.core.routing.controller_stack')->getCurrentController();
        $controller->addJS('myscript.js');
```

So this allows module developer to inject his JS from hook without knowing which controller handled the HTTP request.

## Documentation

As this is for module developers, I will need to write a documentation about it when it's merged


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/19277)
